pkg/controller/direct/alloydb/instance_controller.go

```diff
@@ -0,0 +1,409 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alloydb
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/alloydb/v1beta1"
+	refsv1beta1 "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+
+	gcp "cloud.google.com/go/alloydb/apiv1beta"
+	alloydbpb "cloud.google.com/go/alloydb/apiv1beta/alloydbpb"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func init() {
+	registry.RegisterModel(krm.AlloyDBInstanceGVK, NewInstanceModel)
+}
+
+func NewInstanceModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &instanceModel{config: *config}, nil
+}
+
+var _ directbase.Model = &instanceModel{}
+
+type instanceModel struct {
+	config config.ControllerConfig
+}
+
+func (m *instanceModel) client(ctx context.Context) (*gcp.AlloyDBAdminClient, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewAlloyDBAdminRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("error building AlloyDB client for Instance: %w", err)
+	}
+	return gcpClient, err
+}
+
+func validateRequiredFields(ctx context.Context, reader client.Reader, obj *krm.AlloyDBInstance) error {
+	if obj.Spec.InstanceType != nil && obj.Spec.InstanceTypeRef != nil {
+		return fmt.Errorf("one and only one of 'spec.InstanceTypeRef' " +
+			"and 'spec.InstanceType' should be configured: both are configured")
+	}
+	if obj.Spec.InstanceType == nil && obj.Spec.InstanceTypeRef == nil {
+		return fmt.Errorf("one and only one of 'spec.InstanceTypeRef' " +
+			"and 'spec.InstanceType' should be configured: neither is configured")
+	}
+
+	var instanceType *string
+	if obj.Spec.InstanceType != nil {
+		if *instanceType == "" {
+			return fmt.Errorf("'spec.InstanceType' should be configured with a non-empty string")
+		}
+		instanceType = obj.Spec.InstanceType
+	}
```

Comment:
Is this expected? ln82 will never hit. And as long as spec.instanceType is nil, it always fails.